File: frontend/src/test/fixtures/labor.ts
/**
 * Deterministic test fixtures shaped exactly like the real backend
 * response models (see ../../api/labor.types.ts). Every builder
 * returns a complete, valid object with sensible defaults; tests
 * override only the fields a given scenario cares about. Values here
 * are fixture data, not economic logic -- nothing in this file
 * calculates anything.
 */
import type {
  EmploymentResult,
  EmploymentSectionChanges,
  LaborChangeEvent,
  LaborMonitorResult,
  LaborObservationEvidence,
  LaborWhatChangedResult,
  UnemploymentResult,
  UnemploymentSectionChanges,
} from "../../api/labor.types";

export function buildObservation(overrides: Partial<LaborObservationEvidence> = {}): LaborObservationEvidence {
  return {
    series_id: "PAYEMS",
    observation_date: "2026-07-01",
    value: 157912,
    ...overrides,
  };
}

export function buildEmploymentResult(overrides: Partial<EmploymentResult> = {}): EmploymentResult {
  return {
    series_id: "PAYEMS",
    current_3m_avg_jobs: 150000,
    prior_3m_avg_jobs: 120000,
    momentum_delta_jobs: 30000,
    condition_deadband_jobs: 50000,
    momentum_deadband_jobs: 50000,
    condition: "EXPANDING",
    momentum: "STEADY",
    state: "EXPANDING",
    observations: [
      buildObservation({ observation_date: "2026-01-01" }),
      buildObservation({ observation_date: "2026-02-01" }),
      buildObservation({ observation_date: "2026-03-01" }),
      buildObservation({ observation_date: "2026-04-01" }),
      buildObservation({ observation_date: "2026-05-01" }),
      buildObservation({ observation_date: "2026-06-01" }),
      buildObservation({ observation_date: "2026-07-01" }),
    ],
    ...overrides,
  };
}

export function buildUnemploymentResult(overrides: Partial<UnemploymentResult> = {}): UnemploymentResult {
  return {
    series_id: "UNRATE",
    current_3m_avg: 4.0,
    prior_year_3m_avg: 4.1,
    delta_pp: -0.1,
    unemployment_deadband_pp: 0.2,
    state: "STABLE",
    observations: [
      buildObservation({ series_id: "UNRATE", observation_date: "2026-05-01", value: 4.0 }),
      buildObservation({ series_id: "UNRATE", observation_date: "2026-06-01", value: 4.0 }),
      buildObservation({ series_id: "UNRATE", observation_date: "2026-07-01", value: 4.0 }),
      buildObservation({ series_id: "UNRATE", observation_date: "2025-05-01", value: 4.1 }),
      buildObservation({ series_id: "UNRATE", observation_date: "2025-06-01", value: 4.1 }),
      buildObservation({ series_id: "UNRATE", observation_date: "2025-07-01", value: 4.1 }),
    ],
    ...overrides,
  };
}

export function buildLaborMonitor(overrides: Partial<LaborMonitorResult> = {}): LaborMonitorResult {
  return {
    methodology_id: "labor_v1.0",
    data_basis: "latest_revised_data",
    state: "STRENGTHENING",
    evaluation_period: "2026-07-01",
    employment: buildEmploymentResult(),
    unemployment: buildUnemploymentResult(),
    ...overrides,
  };
}

export function buildLaborChangeEvent(overrides: Partial<LaborChangeEvent> = {}): LaborChangeEvent {
  return {
    component: "EMPLOYMENT",
    event_type: "METRIC_CHANGED",
    field: "current_3m_avg_jobs",
    previous_value: 120000,
    current_value: 150000,
    delta: 30000,
    previous_period: "2026-06-01",
    current_period: "2026-07-01",
    methodology_id: "labor_v1.0",
    data_basis: "latest_revised_data",
    ...overrides,
  };
}

export function buildEmploymentSectionChanges(overrides: Partial<EmploymentSectionChanges> = {}): EmploymentSectionChanges {
  return {
    previous_evidence: buildEmploymentResult(),
    current_evidence: buildEmploymentResult(),
    changes: [],
    state_changed: false,
    metric_changed: false,
    availability_lost: false,
    availability_restored: false,
    ...overrides,
  };
}

export function buildUnemploymentSectionChanges(overrides: Partial<UnemploymentSectionChanges> = {}): UnemploymentSectionChanges {
  return {
    previous_evidence: buildUnemploymentResult(),
    current_evidence: buildUnemploymentResult(),
    changes: [],
    state_changed: false,
    metric_changed: false,
    availability_lost: false,
    availability_restored: false,
    ...overrides,
  };
}

export function buildLaborWhatChanged(overrides: Partial<LaborWhatChangedResult> = {}): LaborWhatChangedResult {
  return {
    methodology_id: "labor_v1.0",
    comparison_contract_id: "labor_what_changed_v1.0",
    comparison_type: "MONTH_OVER_MONTH",
    data_basis: "latest_revised_data",
    comparison_available: true,
    previous_period: "2026-06-01",
    current_period: "2026-07-01",
    previous_labor_state: "STRENGTHENING",
    current_labor_state: "STRENGTHENING",
    employment_changes: buildEmploymentSectionChanges(),
    unemployment_changes: buildUnemploymentSectionChanges(),
    changes: [],
    any_state_changed: false,
    any_metric_changed: false,
    any_availability_changed: false,
    current_labor_result: null,
    ...overrides,
  };
}

/**
 * Published observations for the two Jobs surveys (#51B).
 *
 * `UNRATE` deliberately carries a NULL in the middle, because the live
 * series does (2025-10-01) and because a null coerced to zero is the
 * defect this fixture exists to keep out: it destroys the axis floor
 * AND draws a value the provider never published.
 */
export function buildPayrollObservations(
  overrides: Partial<import("../../api/series.types").SeriesObservationsResponse> = {},
): import("../../api/series.types").SeriesObservationsResponse {
  const observations = [
    { date: "2026-03-01", value: 158600 },
    { date: "2026-04-01", value: 158750 },
    { date: "2026-05-01", value: 158880 },
    { date: "2026-06-01", value: 158960 },
    { date: "2026-07-01", value: 159020 },
    { date: "2026-08-01", value: 159075 },
  ];
  return {
    series_id: "PAYEMS",
    title: "All Employees, Total Nonfarm",
    units: "Thousands of Persons",
    source: "FRED",
    observations,
    pagination: { limit: 200, offset: 0, returned: observations.length, total: observations.length },
    ...overrides,
  };
}

export function buildUnemploymentObservations(
  overrides: Partial<import("../../api/series.types").SeriesObservationsResponse> = {},
): import("../../api/series.types").SeriesObservationsResponse {
  const observations = [
    { date: "2026-03-01", value: 4.3 },
    { date: "2026-04-01", value: 4.2 },
    { date: "2026-05-01", value: null },
    { date: "2026-06-01", value: 4.2 },
    { date: "2026-07-01", value: 4.1 },
    { date: "2026-08-01", value: 4.1 },
  ] as Array<{ date: string; value: number | null }>;
  return {
    series_id: "UNRATE",
    title: "Unemployment Rate",
    units: "Percent",
    source: "FRED",
    observations: observations as Array<{ date: string; value: number }>,
    pagination: { limit: 200, offset: 0, returned: observations.length, total: observations.length },
    ...overrides,
  };
}

File: app/models/labor.py
"""Typed, application-owned canonical result models for the Labor
Market Monitor, methodology `labor_v1.0` (frozen in
`research/labor_momentum/LABOR_V1_FROZEN_METHODOLOGY.md` -- normative;
this module implements it, it does not reinterpret it).

Plain Pydantic data, like `app.models.inflation` -- no FastAPI or
SQLAlchemy dependency. Consumed by `app.domain.labor`'s pure functions
and `app.services.labor`'s orchestration, and returned directly by
`app.api.labor`.

This is the ONE canonical production definition of `labor_v1.0`'s
constants and enums. Nothing else in the application may hardcode a
duplicate series ID or deadband -- every other module imports them
from here.

Deliberately a NEW, independently-defined set of models -- not a
shared base class with `app.models.inflation`'s own
`InflationMetricEvidence`/`SeriesMomentumResult` shapes, even though
the two monitors are structurally similar. This is the second
canonical monitor; per the frozen spec's own "no premature generic
monitor framework" instruction, Labor is implemented independently
first, and any real shared abstraction is deferred until a third
monitor's own requirements can inform what that abstraction should
actually look like.

`data_basis` uses `"latest_revised_data"` -- `app.models.inflation.DATA_BASIS`'s
exact machine-string convention, not the frozen methodology document's
own human-sentence example ("Latest revised data") -- per this
increment's own explicit instruction to use "the existing project
convention" and reconcile in the project's favor when the two differ
(see docs/ENGINEERING_JOURNAL.md's #20B entry for the full account of
this and the other resolved discrepancies).
"""

from datetime import date
from typing import Literal

from pydantic import BaseModel

from app.concepts.bindings import active_binding
from app.models.series import SeriesIdentity

METHODOLOGY_ID = "labor_v1.0"
DATA_BASIS = "latest_revised_data"

# --- Economic concept identity (Increment #38, ADR-034) -------------
#
# `labor_v1.0`'s two components, named by what they measure rather than
# by who publishes them. The distinction between these two concepts is
# not cosmetic: the employment concept counts nonfarm payroll JOBS from
# the CES establishment survey, the unemployment concept counts PEOPLE
# from the CPS household survey. Different universes, different
# surveys, and -- as the #35 Radar research found -- incompatible
# resolutions, which is why neither is ever a substitute for the other.
EMPLOYMENT_CONCEPT_ID = "us.nonfarm.payroll-employment.sa.monthly"
UNEMPLOYMENT_CONCEPT_ID = "us.unemployment-rate.sa.monthly"

# --- Stored series identifiers, DERIVED from the active binding ------
#
# Byte-identical to the literals they replaced; no longer hardcoded.
# These name a place to LOOK IN STORAGE, not MacroChipz's identity.
PAYEMS_SERIES_ID = active_binding(EMPLOYMENT_CONCEPT_ID).storage_series_id
UNRATE_SERIES_ID = active_binding(UNEMPLOYMENT_CONCEPT_ID).storage_series_id

# PAYEMS is persisted in FRED's native "Thousands of Persons" -- every
# `labor_v1.0` formula operates on actual persons/jobs instead (see
# LABOR_V1_FROZEN_METHODOLOGY.md §2). This is the ONE conversion
# constant; `app.domain.labor.build_jobs_index` is the ONE place it is
# applied.
PAYEMS_JOBS_PER_NATIVE_UNIT = active_binding(EMPLOYMENT_CONCEPT_ID).canonical_unit_factor

CONDITION_DEADBAND_JOBS = 50_000
MOMENTUM_DEADBAND_JOBS = 50_000
UNEMPLOYMENT_DEADBAND_PP = 0.2


class LaborSeriesIdentities(BaseModel):
    """Who `labor_v1.0`'s two input series actually are (#38).

    Kept as one object because the two must never be swapped: the
    employment concept counts nonfarm payroll JOBS from the CES
    establishment survey and the unemployment concept counts PEOPLE
    from the CPS household survey. A positional mix-up would attribute
    one survey's evidence to the other.
    """

    employment: SeriesIdentity
    unemployment: SeriesIdentity


EmploymentCondition = Literal["EXPANDING", "FLAT", "CONTRACTING", "INSUFFICIENT_DATA"]
EmploymentMomentum = Literal["IMPROVING", "STEADY", "WORSENING", "INSUFFICIENT_DATA"]
EmploymentState = Literal["EXPANDING", "COOLING", "STABLE", "CONTRACTING", "RECOVERING", "INSUFFICIENT_DATA"]
UnemploymentTrendState = Literal["IMPROVING", "DETERIORATING", "STABLE", "INSUFFICIENT_DATA"]
LaborState = Literal["STRENGTHENING", "COOLING", "STABLE", "MIXED", "INSUFFICIENT_DATA"]


class LaborObservationEvidence(BaseModel):
    """One exact-calendar-month source observation, exactly as
    persisted (FRED-native units -- PAYEMS in thousands of persons,
    UNRATE in percent; the jobs conversion is a pure, deterministic,
    documented multiply, reconstructable from this raw value alone, so
    it is not duplicated here). `value: None` means this exact required
    calendar month has no usable persisted observation -- never
    omitted from the list, so a caller can always see exactly which
    month(s) were missing, not just that something was missing."""

    #: Source-neutral identity (#38), plus the provider that actually
    #: supplied this observation and their own identifier for it. Read
    #: from the persisted series row rather than stamped from a module
    #: constant -- the defect ADR-034 exists to fix.
    concept_id: str
    provider: str
    series_id: str
    observation_date: date
    value: float | None


class EmploymentResult(BaseModel):
    """The PAYEMS-owned half of `labor_v1.0`: employment condition
    (level) and momentum (trend), combined into one `EmploymentState`.
    `condition`/`momentum`/`state` are all `INSUFFICIENT_DATA` together
    whenever any of the 7 exact required PAYEMS calendar months
    (`observations`, `t` through `t-6`) is missing -- never computed
    from a partial window."""

    series_id: str = PAYEMS_SERIES_ID
    current_3m_avg_jobs: float | None
    prior_3m_avg_jobs: float | None
    momentum_delta_jobs: float | None
    condition_deadband_jobs: float = CONDITION_DEADBAND_JOBS
    momentum_deadband_jobs: float = MOMENTUM_DEADBAND_JOBS
    condition: EmploymentCondition
    momentum: EmploymentMomentum
    state: EmploymentState
    observations: list[LaborObservationEvidence]


class UnemploymentResult(BaseModel):
    """The UNRATE-owned half of `labor_v1.0`: current 3-month average
    vs. the 3-month average from exactly one year earlier.
    `INSUFFICIENT_DATA` whenever any of the 6 exact required UNRATE
    calendar months (`observations`) is missing."""

    series_id: str = UNRATE_SERIES_ID
    current_3m_avg: float | None
    prior_year_3m_avg: float | None
    delta_pp: float | None
    unemployment_deadband_pp: float = UNEMPLOYMENT_DEADBAND_PP
    state: UnemploymentTrendState
    observations: list[LaborObservationEvidence]


class LaborMonitorResult(BaseModel):
    """The complete canonical, typed Labor Monitor result -- structured
    evidence, not prose. Every result carries `methodology_id`
    explicitly; it is never inferred from the request path or any
    other context.

    `evaluation_period` is `None` only when neither PAYEMS nor UNRATE
    has any persisted observation at all (no candidate period can be
    chosen) -- distinct from, and never conflated with, `state ==
    "INSUFFICIENT_DATA"` at a genuine, chosen `evaluation_period` where
    one or both components simply lack enough trailing history. Once
    chosen, `evaluation_period` is the SAME calendar month for both
    `employment` and `unemployment` -- this monitor never reports the
    two owners evaluated at two different reference months in one
    result (see LABOR_V1_FROZEN_METHODOLOGY.md §10).
    """

    methodology_id: str = METHODOLOGY_ID
    data_basis: str = DATA_BASIS
    state: LaborState
    evaluation_period: date | None
    employment: EmploymentResult
    unemployment: UnemploymentResult
